### DISCUSSION TITLE
treeTable多样化编辑问题

### DISCUSSION THREAD

**Discussion by hjingsuper** [layui/layui#1774] [讨论]
示例：https://codepen.io/hjingsuper/pen/MWRGrrr

> 1.treeTable中使用 原生 select，多次点击select会导致节点展开或关闭

![2024-04-10-10-37-04](https://github.com/layui/layui/assets/36070786/1767bd80-d731-409a-b919-2e40f4685b8a)

> 2.使用table的多样化编辑 原生 select 模式，treeTable好像没有cache导致错误
最终 `treeTable.getData(id, isSimpleData);` 获取不到下拉后的值

```
  var options = this;
  
  // 获取当前行数据
  table.getRowData = function(tableId, elem){
    var index = $(elem).closest('tr').data('index');
    return table.cache[tableId][index] || {};
  };
  
  // 原生 select 事件
  var tableViewElem = this.elem.next();
  tableViewElem.find('.select-demo-primary').on('change', function(){
    var value = this.value; // 获取选中项 value
    //treeTable这样写会报错
    var data = table.getRowData(options.id, this); // 获取当前行数据(如 id 等字段，以作为数据修改的索引)
    // 更新数据中对应的字段
    data.city = value;
    // 显示 - 仅用于演示
    layer.msg('选中值: '+ value +'<br>当前行数据：'+ JSON.stringify(data));
  });
```

**Comment by Sight-wcg** [comment]
> 1.treeTable中使用 原生 select，多次点击select会导致节点展开或关闭

添加设置
```
    tree: {
      view: {
        dblClickExpand: false,  // 双击节点时，是否自动展开父节点
      }
    }
```

> 2. 使用table的多样化编辑 原生 select 模式，treeTable好像没有cache导致错误
最终 treeTable.getData(id, isSimpleData); 获取不到下拉后的值

按照 codepen 中的示例，无法重现

**Reply by hjingsuper** [reply]
> 1.treeTable中使用 原生 select，多次点击select会导致节点展开或关闭

已解决，谢谢

> 2.使用table的多样化编辑 原生 select 模式，treeTable好像没有cache导致错误

不好意思，codepen 已修正
新增，编辑用户名，性别。保存，只有用户名字段
![image](https://github.com/layui/layui/assets/36070786/6ab4f07c-9018-4e54-9861-26f9d5a2750f)

**Comment by Sight-wcg** [comment]
问题描述中的代码是正确的，codepen 中的错误，只能使用 table.cache
> table.getRowData = function(tableId, elem){
>     var index = $(elem).closest('tr').data('index');
>     return table.cache[tableId][index] || {};
>   };

```diff
  table.getRowData = function(tableId, elem){
    var index = $(elem).closest('tr').data('index');
-    return treeTable.cache[tableId][index] || {};
+    return layui.table.cache[tableId][index] || {};
  };
```

**Reply by hjingsuper** [reply]
示例 [https://codepen.io/hjingsuper/pen/MWRGrrr](https://codepen.io/hjingsuper/pen/MWRGrrr) 已更新，不知道对不对，麻烦大佬再看下
> 1.意思是treeTable的多样化编辑，只能使用table.cache？这样很奇怪，不知道后期会不会做修改，这个要确定下

> 2.这样好像有bug，第一次选择后会赋值，之后不会赋值


![image](https://github.com/layui/layui/assets/36070786/33ec4bb6-4c85-40dc-b7a6-55560b88fe74)

**Reply by Sight-wcg** [reply]
> 1.意思是treeTable的多样化编辑，只能使用table.cache？这样很奇怪，不知道后期会不会做修改，这个要确定下

是的，只能使用 table.cache，应该是设计如此，不会改了
> 2.这样好像有bug，第一次选择后会赋值，之后不会赋值

我无法重现这个问题，有没有详细的操作步骤
![image](https://github.com/layui/layui/assets/26325820/dfd4b2ab-2eec-4cb3-b5d7-b0de8541e74a)

**Reply by hjingsuper** [reply]
> 1.意思是treeTable的多样化编辑，只能使用table.cache？这样很奇怪，不知道后期会不会做修改，这个要确定下

ok，谢谢确认

> 2.这样好像有bug，第一次选择后会赋值，之后不会赋值

**确定示例layui.js引用无问题、代码无问题情况下**

发现和操作顺序无关，点击【保存】时 `treeTable.getData("MianTable")` 只取到第一个行下拉值
不知道您怎么操作的，能否发出来看下🙏

**第一种操作：看log，只有【用户名 1】有【性别 男】的值**

1. 新增主件 -> 输入【用户名 1】、选择【性别 男】
2. 添加【1】的子件 -> 输入【用户名 2】、选择【性别 女】
3. 新增主件 -> 输入【用户名 3】、选择【性别 男】

![2024-04-10-15-02-41](https://github.com/layui/layui/assets/36070786/8a19d44a-469a-4f5d-8fb4-be58f7e87fa3)
![2024-04-10-15-03-02](https://github.com/layui/layui/assets/36070786/ac74b5c3-e799-4c8a-8c33-0a1c83492137)

**第二种操作：看log，只有【用户名 1】有【性别 男】的值**

1. 新增主件 -> 输入【用户名 1】、选择【性别 男】
2. 新增主件 -> 输入【用户名 3】、选择【性别 男】
3. 添加【3】的子件 -> 输入【用户名 2】、选择【性别 女】

![2024-04-10-15-07-30](https://github.com/layui/layui/assets/36070786/c8c5fc1c-41a1-4be2-ae4d-6b43a5f3d1b3)
![2024-04-10-15-07-51](https://github.com/layui/layui/assets/36070786/0f0b99a9-3392-484e-abe1-47b38f86ae29)

**第三种操作：看log，只有【用户名 1】有【性别 女】的值**

1. 新增主件
2. 新增主件 -> 添加的子件 -> 子件 输入【用户名 3】、选择【性别 男】 -> 父级 输入【用户名 2】、选择【性别 女】
3. 最后第一个主件输入【用户名 1】、选择【性别 女】

![2024-04-10-15-17-39](https://github.com/layui/layui/assets/36070786/0732e46f-5b99-46e6-b10c-864c609db593)
![2024-04-10-15-18-18](https://github.com/layui/layui/assets/36070786/b7e5daab-6f88-47e6-a909-38382948f556)

**Reply by Sight-wcg** [reply]
> 发现和操作顺序无关，点击【保存】时 treeTable.getData("MianTable") 只取到第一个行下拉值
不知道您怎么操作的，能否发出来看下

我是点击 “初始化按钮” > 修改三行单元格内容  >  点击"保存按钮"，所以没能重现

> 第一次选择后会赋值，之后不会赋值

treeTable 内部添加新行时，如果查找到 “无数据” 界面相关的 class，会直接 reload，reload 完成之后，触发 done 回调，执行里面的代码，为 select 元素添加 change 事件侦听器。所以操作第一行的 select 元素, 可以正常更新 tableCache。

而后续点击 “新增主件” 添加新行，并不会触发 done 回调，其中的 select 元素没有添加 change 事件侦听器，对其操作也就无法更新 tableCache。

这个场景下涉及到动态增删 DOM，推荐使用事件委托，以下仅供参考
```
      var tableViewElem = this.elem.next();
      // 解除 tbSelect 命名空间下的所有 change 事件处理程序
      tableViewElem.off("change.tbSelect");
      // 将 '.select-sex' 元素的 change 事件委托给 tableViewElem, 事件命名空间为 tbSelect
      tableViewElem.on("change.tbSelect", ".select-sex", function () {
        var value = this.value; // 获取选中项 value
        var data = table.getRowData(options.id, this);
        data.sex = value;
      });
      // 将 '.select-province' 元素的 change 事件委托给 tableViewElem, 事件命名空间为 tbSelect
      tableViewElem.on("change.tbSelect", ".select-province", function () {
        var value = this.value; // 获取选中项 value
        var data = table.getRowData(options.id, this);
        data.city = value;
      });
```

目前单元格编辑用起来不太方便，需要注意的事项比较多，例如更新/添加行之后事件需要手动重新绑定，要彻底解决此类问题，需要 templet 支持将 DOM 对象 append 到 td 元素中，而不是使用 innerHtml。我在本地有尝试过改善这一点，但在我的设备上，表格性能下降了 20% 左右（和处理器，系统，浏览器有关）。我目前在性能优化方面没什么经验，只能暂时搁置了，将来也许会改进这一块。
![image](https://github.com/layui/layui/assets/26325820/7f9e8583-2c23-4cc7-a789-bcc4ead0b824)

**Reply by hjingsuper** [reply]
使用事件委托，问题得到解决。谢谢🫰
性能问题，暂时不必担心。后期会提供导入功能，提升效率
再次感谢
